centralizedmetrics/metrics.go

```diff
@@ -0,0 +1,139 @@
+package centralizedmetrics
+
+import (
+	"database/sql"
+	"sync"
+	"time"
+
+	"github.com/ethereum/go-ethereum/log"
+
+	"github.com/status-im/status-go/centralizedmetrics/common"
+	"github.com/status-im/status-go/centralizedmetrics/providers"
+)
+
+var defaultPollInterval = 10 * time.Second
+
+type MetricsInfo struct {
+	Enabled       bool `json:"enabled"`
+	UserConfirmed bool `json:"userConfirmed"`
+}
+
+type MetricRepository interface {
+	Poll() ([]common.Metric, error)
+	Delete(metrics []common.Metric) error
+	Add(metric common.Metric) error
+	Info() (*MetricsInfo, error)
+	ToggleEnabled(enabled bool) error
+}
+
+type MetricService struct {
+	repository MetricRepository
+	processor  common.MetricProcessor
+	ticker     *time.Ticker
+	done       chan bool
+	started    bool
+	wg         sync.WaitGroup
+}
+
+func NewDefaultMetricService(db *sql.DB) *MetricService {
+	repository := NewSQLiteMetricRepository(db)
+	processor := providers.NewMixpanelMetricProcessor(providers.MixpanelAppID, providers.MixpanelToken, providers.MixpanelBaseURL)
+	return NewMetricService(repository, processor, defaultPollInterval)
+}
+
+func NewMetricService(repository MetricRepository, processor common.MetricProcessor, interval time.Duration) *MetricService {
+	return &MetricService{
+		repository: repository,
+		processor:  processor,
+		ticker:     time.NewTicker(interval),
+		done:       make(chan bool),
+	}
+}
+
+func (s *MetricService) Start() {
+	if s.started {
+		return
+	}
+	s.wg.Add(1)
+	s.started = true
+	go func() {
+		defer s.wg.Done()
+		for {
+			select {
+			case <-s.done:
+				return
+			case <-s.ticker.C:
+				s.processMetrics()
+			}
+		}
+	}()
+}
+
+func (s *MetricService) Stop() {
+	if !s.started {
+		return
+	}
+	s.ticker.Stop()
+	s.done <- true
+	s.wg.Wait()
+	s.started = false
+}
+
+func (s *MetricService) EnsureStarted() error {
+	info, err := s.Info()
+	if err != nil {
+		return err
+	}
+	if info.Enabled {
+		s.Start()
+	}
+	return nil
+}
+
+func (s *MetricService) Info() (*MetricsInfo, error) {
+	return s.repository.Info()
+}
+
+func (s *MetricService) ToggleEnabled(enabled bool) error {
+	err := s.repository.ToggleEnabled(enabled)
+	if err != nil {
+		return err
+	}
+	if enabled {
+		s.Start()
+	} else {
+		s.Stop()
+	}
+	return nil
+
+}
+
+func (s *MetricService) AddMetric(metric common.Metric) error {
+	return s.repository.Add(metric)
+}
+
+func (s *MetricService) processMetrics() {
+	log.Info("processing metrics")
+	metrics, err := s.repository.Poll()
+	if err != nil {
+		log.Warn("error polling metrics", "error", err)
+		return
+	}
+	log.Info("polled metrics")
+
+	if len(metrics) == 0 {
+		return
+	}
+	log.Info("processing metrics")
+
+	if err := s.processor.Process(metrics); err != nil {
```

Comment:
It looks like `Process` short circuits on the first failure to send a metric and then all the remaining metrics are deleted. I'm just wondering, an HTTP request failure can be very common and there may be better ways to handle failure.

For example, skip failed ones and process the remaining in the batch of 10 we pulled from the db. Or depending on the HTTP failure, like a 503, it would be probably better to wait a little while before sending metrics to the provider again otherwise many will fail in succession. Or if the token is rate limited, the providers probably reply in a specific way, so it would be important to detect that and stop sending metrics to not make things worse. For that it would be really nice to have a form of rate limiter and I don't know if status-go already uses/has something like this.

We process one token per second correct? Is that relatively safe to prevent the API from rate limiting us based on the amount of users we expect to get from 2.30? Probably yes, right? Because we're not going to track too many events now.

It's complicated and a problem we currently don't have for the initial release, but as we introduce more metrics in the app I think this sort of problem will creep up, even without malicious users 😈

Reply:
I haven't made this code robust yet, that would be a separate PR, they give us specific feedback on how to handle retries etc, but I would do that in a separate PR. 
One thing, I think if one of the metrics fails, they won't be deleted (which also is an issue). But I would address in a separate issue, I think we can do it for 2.30, but in a scope of a separate PR, would that be ok?

Reply:
Of course, I missed the early return before deletion, sorry about that.

> I think we can do it for 2.30, but in a scope of a separate PR, would that be ok?

For sure @cammellos, I wouldn't recommend this for this PR 👍🏼 

> they give us specific feedback on how to handle retries etc, but I would do that in a separate PR.

Do they recommend as well how to handle rate limiting? Are you considering also minimizing this problem for 2.30?

Reply:
they do, rate limit is very generous, but still we can follow their guidelines